armory/scenarios/video_ucf101_scenario_adversarial_preloaded_and_defended.py

```diff
@@ -0,0 +1,215 @@
+                defense = load(config.get("defense"))
+            else:
+                logger.info(f"Applying internal {defense_type} defense to classifier")
+                classifier = load_defense_internal(config["defense"], classifier)
+
+        if model_config["fit"]:
+            classifier.set_learning_phase(True)
+            logger.info(
+                f"Fitting model {model_config['module']}.{model_config['name']}..."
+            )
+            train_epochs = config["model"]["fit_kwargs"]["nb_epochs"]
+            batch_size = config["dataset"]["batch_size"]
+
+            logger.info(f"Loading train dataset {config['dataset']['name']}...")
+            train_data = load_dataset(
+                config["dataset"],
+                epochs=train_epochs,
+                split_type="train",
+                preprocessing_fn=preprocessing_fn,
+            )
+
+            if defense_type == "Trainer":
+                logger.info(f"Training with {defense_type} defense...")
+                defense = load_defense_wrapper(config["defense"], classifier)
+            else:
+                logger.info(f"Fitting classifier on clean train dataset...")
+
+            for epoch in range(train_epochs):
+                classifier.set_learning_phase(True)
+
+                for _ in tqdm(
+                    range(train_data.batches_per_epoch),
+                    desc=f"Epoch: {epoch}/{train_epochs}",
+                ):
+                    x, y = train_data.get_batch()
+                    # x_trains consists of one or more videos, each represented as an
+                    # ndarray of shape (n_stacks, 3, 16, 112, 112).
+                    # To train, randomly sample a batch of stacks
+                    x = np.stack([x_i[np.random.randint(x_i.shape[0])] for x_i in x])
+                    if defense_type == "Trainer":
+                        defense.fit(x, y, batch_size=batch_size, nb_epochs=1)
+                    else:
+                        # apply defense outside of classifier
+                        if IMAGE_DEFENSE_SUPPORT_ONLY_4D:
+                            for ns in range(x.shape[0]):
+                                xt = np.transpose(x[ns, ...], (1, 2, 3, 0))
+                                if defense_config.get("name") == "JpegCompression":
+                                    # jpegcompression requires unnormalized inputs
+                                    xt[..., 0] += 114.7748
+                                    xt[..., 1] += 107.7354
+                                    xt[..., 2] += 99.4750
+                                xt, _ = defense(xt)
+                                if defense_config.get("name") == "JpegCompression":
+                                    xt[..., 0] -= 114.7748
+                                    xt[..., 1] -= 107.7354
+                                    xt[..., 2] -= 99.4750
+                                x[ns, ...] = np.transpose(xt, (3, 0, 1, 2))
```

Comment:
This shows up several places in the script. Can you abstract this into a Preprocessing ART function in armory.art_experimental.defences that wraps the underlying JpegCompression and works seamlessly in the original scenario?

Ideally, the hard-coded means and transpose args would be stored in the defense's `kwargs` portion of the config.

Reply:
If you can point me to an example of how an experimental defense is used in a scenario, then I can try to take a crack at it if the effort isn't too large.  Otherwise, I'm leaving the ART portion to the IBM team and going to focus on scenarios instead.

Reply:
I tested this out and it seemed to work. This is my file `armory.art_experimental.defences.jpeg_compression_5d.py`:
```
from art.defences import JpegCompression                                                 
import numpy as np                                                                       
                                                                                         
                                                                                         
class JpegCompression5D(JpegCompression):                                                
    def __init__(                                                                        
        self,                                                                            
        clip_values,                                                                     
        quality=50,                                                                      
        channel_index=3,                                                                 
        apply_fit=True,                                                                  
        apply_predict=False,                                                             
        transpose=None,                                                                  
        means=None,                                                                      
    ):                                                                                   
        super().__init__(                                                                
            clip_values,                                                                 
            quality=quality,                                                             
            channel_index=channel_index,                                                 
            apply_fit=apply_fit,                                                         
            apply_predict=apply_predict,                                                 
        )                                                                                
        # transpose = (1, 2, 3, 0),                                                      
        if transpose is None:                                                            
            transpose = (0, 1, 2, 3)  # identify operation                               
        if len(transpose) != 4:                                                          
            raise ValueError("transpose must have 4 dimensions specified")               
        self.transpose = transpose                                                       
                                                                                         
        # means = (114.7748, 107.7354, 99.4750),                                         
        if means is None:                                                                
            means = (0.0, 0.0, 0.0)  # identity operation                                
        if len(means) != 3:                                                              
            raise ValueError("means must have 3 values, one per channel")                
        self.means = means                                                               
                                                                                         
    def __call__(self, x, y=None):                                                       
        x_new = []                                                                       
        for x_i in x:                                                                    
            x_i = x_i.transpose(self.transpose)                                          
            x_i = x_i + self.means  # use broadcasting                                   
            x_i, _ = super().__call__(x_i)                                               
            x_i = x_i - self.means                                                       
            x_i = x_i.transpose(np.argsort(self.transpose))                              
            x_new.append(x_i)                                                            
        return np.stack(x_new), y 
```

To test:
```
from armory.art_experimental.defences import jpeg_compression_5d
defense = jpeg_compression_5d.JpegCompression5D(clip_values=(0, 255))
ds = datasets.ucf101()
defense(x)
```
For use with the MARS preprocessing function:
```
from armory.art_experimental.defences import jpeg_compression_5d
defense = jpeg_compression_5d.JpegCompression5D(clip_values=(0, 255), tranpose=(1, 2, 3, 0), means=(114.7748, 107.7354, 99.4750))
ds = datasets.ucf101(preprocessing_fn=preprocessing_fn)
defense(x)
```

Reply:
I think that if you add this file and then modify your config, this should work with the original scenario, but let me know if you run into issues.